source/Cache.h

```diff
@@ -0,0 +1,326 @@
+/* Cache.h
+Copyright (c) 2018 by OOTA, Masato
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
+
+#ifndef CACHE_H_
+#define CACHE_H_
+
+#include <cstddef>
+#include <functional>
+#include <list>
+#include <unordered_map>
+#include <utility>
+
+
+
+// This is the common timer class for the Cache class.
+class CacheBase {
+public:
+	CacheBase() = default;
+	virtual ~CacheBase() = default;
+	
+	// Set an interval to change the generations.
+	// An auto expired cache may recycle a value after changing 2 generations.
+	// The newInterval's unit is a number of steps.
+	void SetUpdateInterval(std::size_t newInterval);
+	
+	// Notify to progress a frame time.
+	// This function will call StepThis() for all instances of Cache.
+	static void Step();
+	
+protected:
+	CacheBase &operator=(CacheBase &a) noexcept = default;
+	
+	// Register an instance of CacheBase when constructing it.
+	static void RegisterCacheObject(CacheBase *cacheObject);
+	// Unregister an instance of CacheBase when destructing it.
+	static void UnregisterCacheObject(CacheBase *cacheObject);
+	
+private:
+	// Called from Step().
+	void StepThis();
+	// Notify to progress a generation.
+	virtual void NextGeneration() = 0;
+	
+	std::size_t stepCount = 0;
+	std::size_t updateInterval = 3600;
+};
+
+
+
+// Do nothing at recycle of a cache element.
+template<class T>
+class DoNothingAtRecycle {
+public:
+	void operator()(T& data) const {};
+};
+
+
+
+// This is a generic cache class which has some template parameters same as unordered_map.
+// Key is a type of the key values.
+// T is a type of the mapped values.
+// Hash is a type of the unary function that returns the hash value of a key value.
+// If autoExpired is true, its instance may recycle a value after changing 2 generations. It means that any data keep for one generation at least.
+// If autoExpired is false, the cache keeps all data unless Expire() has been called.
+// AtRecycle is a function class. The function is called before recycling a value.
+template<class Key, class T, bool autoExpired = false, class Hash = std::hash<Key>, class AtRecycle = DoNothingAtRecycle<T>>
```

Comment:
Do we have a use in the game's context where a manually managed cache makes sense? If not, should we remove the "autoExpired" template parameter, implementing only an automatically managed cache?

Reply:
Probably, no, and yes.